modules/vault-cluster/main.tf

```diff
@@ -72,6 +111,77 @@ resource "aws_launch_configuration" "launch_configuration" {
   }
 }
 
+# ---------------------------------------------------------------------------------------------------------------------
+# CREATE LAUNCH TEMPLATE TO DEFINE WHAT RUNS ON EACH INSTANCE IN THE ASG
+# ---------------------------------------------------------------------------------------------------------------------
+
+data "aws_ami" "ami" {
+  filter {
+    name   = "image-id"
+    values = ["${var.ami_id}"]
+  }
+}
+
+resource "aws_launch_template" "launch_template" {
+  count         = "${var.asg_launch_mechanism == "launch_template" ? 1 : 0}"
+  name_prefix   = "${var.cluster_name}-"
+  image_id      = "${var.ami_id}"
+  instance_type = "${var.instance_type}"
+  user_data     = "${base64encode(var.user_data)}"
+
+  depends_on = ["aws_iam_instance_profile.instance_profile"]
+
+  iam_instance_profile {
+    name = "${aws_iam_instance_profile.instance_profile.name}"
+  }
+
+  key_name = "${var.ssh_key_name}"
+
+  # Don't use vpc_security_group_ids when network_interfaces includes security_groups.
+  # vpc_security_group_ids = ["${concat(list(aws_security_group.lc_security_group.id), var.additional_security_group_ids)}"]
+
+  placement {
+    tenancy = "${var.tenancy}"
+  }
+  network_interfaces {
+    associate_public_ip_address = "${var.associate_public_ip_address}"
+    delete_on_termination       = true
+    security_groups             = ["${concat(list(aws_security_group.lc_security_group.id), var.additional_security_group_ids)}"]
+  }
+  ebs_optimized = "${var.root_volume_ebs_optimized}"
+  block_device_mappings {
+    device_name = "${data.aws_ami.ami.root_device_name}"
+
+    ebs {
+      encrypted             = "${var.root_volume_ebs_encryption}"
+      volume_type           = "${var.root_volume_type}"
+      volume_size           = "${var.root_volume_size}"
+      delete_on_termination = "${var.root_volume_delete_on_termination}"
+    }
+  }
+  tags = "${var.launch_template_tags}"
+  tag_specifications {
+    # Instance tags are already handled by the autoscaling group
+    resource_type = "volume"
```

Comment:
I must admit that this is a lot of extra code to maintain just to be able to tag EBS volumes. Is there no other way to do it?

Reply:
I agree. My hope is that others will find other reasons why launch_template is a better alternative for their usecases and can expand on what I've started here.

The reason for the long delay between my initial submission & issue and this PR was due to the aws provider issue mentioned elsewhere. In the meantime, I was able to find a workaround via user-data script which inspects the instance & applies its tags to the instance volume. Of course, that requires that the instance be able to query its own tags and apply tags to its instance.

So, yes, there is another way to do it but it's a hack. And, from what I gather, launch templates are the new hot thing and presumably the preferred option going forward.